### PR TITLE
Added a UI for viewing the religions inside a city

### PR DESCRIPTION
**Pull request by xlenstra**
This PR adds a small UI for viewing the followers of religions inside a city.

<details> <summary> Picture: </summary>

image

</details>

Feedback is welcome.

Note that this PR will have merge conflicts with both #4902 and #4890.

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
- Takes lots of space, maybe too much. I just considered using an ExpanderTab in there folding away the specialists and thought 'not really worth it'... Also, don't know how flexible I made them - whether one can do a really small one that wouldn't blow up the 'Majority Religion' line much when used as its title... Also, would need the stuff I'm working on now (temporary persistence - got derailed as usual).
- First user-visible use of "Pressure"? Decide now once and for all whether that's the UI term we want for that? Sounds a little constipated... Or is it perfectly appropriate because the phenomenon is comparable to the mafia exerting pressure on a shop owner?
- I wouldn't show "pressure" numbers, just a kinda progress bar style widget. Maybe a nice graphic partially greyed - flames with crosses? Without the burning witches? Could be a little challenge whipping up such a widget.

That reminds me - the game atlas is again very close to overflowing. Someone should test whether halving the pixel dimensions of the pure 'Stat' sprites wouldn't hurt...

**Comment by xlenstra** [comment]
For comparison, this is how the equivalent table in the original looks (image shamelessly stolen from google search):

<details>

image

</details>

**Comment by xlenstra** [comment]
>     * Takes lots of space, maybe too much. I just considered using an ExpanderTab in there folding away the specialists and thought 'not really worth it'... Also, don't know how flexible I made them - whether one can do a really small one that wouldn't blow up the 'Majority Religion' line much when used as its title... Also, would need the stuff I'm working on now (temporary persistence - got derailed as usual).

ExpanderTabs do sound like a good idea, I'll look into them

>     * First user-visible use of "Pressure"? Decide now once and for all whether that's the UI term we want for that? Sounds a little constipated... Or is it perfectly appropriate because the phenomenon is comparable to the mafia exerting pressure on a shop owner?

It is the term used by the original, if we want to deviate from that we can, of course, but I don't have any great ideas myself so unless someone else does, I'm inclined to keep the original term

>     * I wouldn't show "pressure" numbers, just a kinda progress bar style widget. Maybe a nice graphic partially greyed - flames with crosses? Without the burning witches? Could be a little challenge whipping up such a widget.

Progress bar style widget is difficult, as the pressure of each religion is effectively racing against the pressures of all other religions. After each turn the pressure shown is added to a hidden pool of pressure, and followers are then allocated approximately in the same proportions as the hidden pools of pressure. Using a progress bar style widget could in some cases therefore result in the progress bar getting more empty as time passes, which is quite counterintuitive for a progress bar.

**Comment by SomeTroglodyte** [comment]
OK, so "Pressure" :hankey:   it is...
Hey look . the religion symbol on the city panel has shadows! Need those too now. I had an Unciv branch with working _text_ shadows once...

> ExpanderTabs

... are what I'm expanding now (extra callback for dynamic resizing upstream). Header fontsize is parameterized but, header background and padding are not accessible ... and 10f padding is too much in this place. Arrow size static too. Wrapping and applying setScale would likely work, but - ugly.

> counterintuitive for a progress bar

Not worse than counterintuitive numbers? Nah, I'll shut up. I haven't researched these mechanics => unqualified. But - `public class Slider extends ProgressBar` something in that direction must already exist in Gdx... Wonder why our current progress-style thingies aren't using it. Probably because ours show two values...

**Comment by ravignir** [comment]
> For comparison, this is how the equivalent table in the original looks (image shamelessly stolen from google search):
> 
> image
>
This image shows one more thing: pressure is visible to other civs as well.

**Comment by SomeTroglodyte** [comment]
Sorry for the off-topic, but I think it's enough you know about it, not worth an issue: Load game -> CityInfoReligionManager.setTransients -> updateNumberOfFollowers -> triggerReligionAdoption -> infinite 'was converted to' messages

**Comment by SomeTroglodyte** [comment]
Getting that to look nice with an ExpanderTab has some pitfalls... Can't capture video here, so if you want, just try it. Remember, this keeps ExpanderTab's static formatting intact, so no tweaking its background or header padding in this one.

<details><summary>patch (on top of your branch with merged-in current master):</summary>

```patch
Index: core/src/com/unciv/ui/cityscreen/CityStatsTable.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/src/com/unciv/ui/cityscreen/CityStatsTable.kt b/core/src/com/unciv/ui/cityscreen/CityStatsTable.kt
--- a/core/src/com/unciv/ui/cityscreen/CityStatsTable.kt	(revision f2e3e9baafc24c129a90c6744203ec0177aa564e)
+++ b/core/src/com/unciv/ui/cityscreen/CityStatsTable.kt	(date 1629373974513)
@@ -2,7 +2,9 @@
 
 import com.badlogic.gdx.graphics.Color
 import com.badlogic.gdx.scenes.scene2d.ui.Table
+import com.badlogic.gdx.utils.Align
 import com.unciv.logic.civilization.ReligionState
+import com.unciv.models.UncivSound
 import com.unciv.models.stats.Stat
 import com.unciv.models.translations.tr
 import com.unciv.ui.utils.*
@@ -40,7 +42,7 @@
         addText()
         if (!cityInfo.population.getMaxSpecialists().isEmpty()) {
             innerTable.addSeparator()
-            innerTable.add(SpecialistAllocationTable(cityScreen).apply { update() })
+            innerTable.add(SpecialistAllocationTable(cityScreen).apply { update() }).row()
         }
 
         if (cityInfo.religion.getNumberOfFollowers().isNotEmpty())
@@ -87,42 +89,50 @@
         innerTable.addSeparator()
         val label = cityInfo.religion.getMajorityReligion()
             ?: "None"
-        innerTable.add("Majority Religion: [$label]".toLabel()).pad(5f).row()
-        // This line above will have merge conflicts with #4902, this version should have priority
-
-        innerTable.add(addReligionsTable())
+        innerTable.add(ExpanderTab(
+            title = "Majority Religion: [$label]",
+            fontSize = 18,
+            icon = ImageGetter.getReligionIcon("Religion").surroundWithCircle(30f, color=Color()),
+            defaultPad = 0f,
+            persistenceID = "CityStatsTable.Religion",
+            onChange = {
+                Sounds.play(UncivSound.Choir)   // small joke here
+                pack()
+                // re-anchor. That alignment instruction was not kept.
+                setPosition(stage.width - CityScreen.posFromEdge, stage.height - CityScreen.posFromEdge, Align.topRight)
+            }
+        ){
+            fillReligionsTable(it)
+        }).growX().row()
     }
-
-    private fun addReligionsTable(): Table {
-        val religionsTable = Table(CameraStageBaseScreen.skin)
+    
+    private fun fillReligionsTable(religionsTable: Table) {
+        val gridColor = Color.DARK_GRAY
         val followers = cityInfo.religion.getNumberOfFollowers()
         val futurePressures = cityInfo.religion.getPressuresFromSurroundingCities()
 
-        religionsTable.add("Religion").pad(5f)
-        religionsTable.addSeparatorVertical()
+        religionsTable.add("Religion".toLabel()).pad(5f)
+        religionsTable.addSeparatorVertical(gridColor)
         val followerImage = ImageGetter.getReligionIcon("Follower")
             .apply { color = Color.BLACK }
         religionsTable.add(followerImage.surroundWithCircle(30f)).pad(5f)
-        religionsTable.addSeparatorVertical()
-        religionsTable.add("Pressure").pad(5f)
+        religionsTable.addSeparatorVertical(gridColor)
+        religionsTable.add("Pressure".toLabel()).pad(5f)
         religionsTable.row()
-        religionsTable.addSeparator()
+        religionsTable.addSeparator(gridColor)
 
         for ((religion, followerCount) in followers) {
             val religionImage = ImageGetter.getReligionIcon(cityInfo.civInfo.gameInfo.religions[religion]!!.iconName)
                 .apply { color = Color.BLACK }
             religionsTable.add(religionImage.surroundWithCircle(30f)).pad(5f)
-            religionsTable.addSeparatorVertical()
+            religionsTable.addSeparatorVertical(gridColor)
             religionsTable.add(followerCount.toLabel()).pad(5f)
-            religionsTable.addSeparatorVertical()
+            religionsTable.addSeparatorVertical(gridColor)
             if (futurePressures.containsKey(religion))
                 religionsTable.add(("+ [${futurePressures[religion]!!}] pressure").toLabel()).pad(5f)
             else
                 religionsTable.add()
             religionsTable.row()
         }
-
-
-        return religionsTable
     }
-}
\ No newline at end of file
+}
```
</details>

File: core/src/com/unciv/logic/city/CityReligion.kt [hunk 1]
```diff
@@ -214,4 +215,24 @@ class CityInfoReligionManager {
         
         updateNumberOfFollowers()
     }
+    
+    /** Doesn't update the pressures, only returns what they are if the update were to happen right now */
+    fun getPressuresFromSurroundingCities(): Counter<String> {
+        val addedPressure = Counter<String>()
+        if (cityInfo.isHolyCity()) {
+            addedPressure[religionThisIsTheHolyCityOf!!] = 5 * pressureFromAdjacentCities
+        }
```
Comment (yairm210) [inline core/src/com/unciv/logic/city/CityReligion.kt:224]: I highly doubt this is how it works. If I'm a holy city I have 5x the *existing* pressure as pressure *added*?

File: core/src/com/unciv/logic/city/CityReligion.kt [hunk 2]
```diff
@@ -214,4 +215,24 @@ class CityInfoReligionManager {
...
+        val allCitiesWithin10Tiles =
+            cityInfo.civInfo.gameInfo.getCities()
+                .filter {
+                    it != cityInfo
+                            && it.getCenterTile().aerialDistanceTo(cityInfo.getCenterTile()) <= 10
+                }
+        for (city in allCitiesWithin10Tiles) {
+            val majorityReligionOfCity = city.religion.getMajorityReligion() ?: continue
+            if (!cityInfo.civInfo.gameInfo.religions[majorityReligionOfCity]!!.isMajorReligion()) continue
```
Comment (yairm210) [inline core/src/com/unciv/logic/city/CityReligion.kt:233]: I think there should be a getMajorityReligionObject() function in this class, seems like something that has many uses

File: core/src/com/unciv/logic/city/CityReligion.kt (hunk 1 above)
Comment (xlenstra) [reply core/src/com/unciv/logic/city/CityReligion.kt:224]: `pressureFromAdjacentCities` is a cached variable of the pressure a single city adds to adjacent cities (excluding itself). Holy cities do pressure themselves, and 5 times as much as adjacent cities pressure each other, so this is implemented correctly

Comment (yairm210) [reply core/src/com/unciv/logic/city/CityReligion.kt:224]: Huh
Ok then

**Comment by xlenstra** [comment]
Using your code I've updated the UI, so new picture:

<details>

image

</details>

**Comment by logicminimal** [comment]
> Using your code I've updated the UI, so new picture:
> 
> <details>
> 
> image
> 
> </details>

does that ui, by any chance show the name of that pantheon and religion for when custom religion names be implemented?

**Comment by xlenstra** [comment]
> > Using your code I've updated the UI, so new picture:
> > 
> > image
> 
> does that ui, by any chance show the name of that pantheon and religion for when the custom religion names are to be implemented?

Currently no, but that might be implemented when custom religion names are implemented.

**Comment by SomeTroglodyte** [comment]
> Using your code I've updated the UI

Aww, those folded hands are so pretty and we see them so rarely :wink:  But - any critique on the ExpanderTab API? Where would you have had most trouble using it, what's counter-intuitive?

**Comment by yairm210** [comment]
Looks like we have some... * puts on sunglasses * religious conflict.

**Comment by SomeTroglodyte** [comment]
Then we need a flame war... Sorry. Must come from #4890 . 🕶️🙏🏽

**Comment by xlenstra** [comment]
> > Using your code I've updated the UI
> 
> Aww, those folded hands are so pretty and we see them so rarely wink But - any critique on the ExpanderTab API? Where would you have had most trouble using it, what's counter-intuitive?

I tinkered around with it a bit before just copy+pasting the parts of your excerpt that looked useful to get it working. API is fine, only thing that felt weird is that the contents of the `ExpanderTab` should be in a `{}` block after the initialization of the `ExpanderTab`, but that might just be general Koitlin weirdness. For me it feels more natural to either pass it as a parameter (similar to the `action` value of `UnitAction`s) or later add it in with an `add()` function (similar to tables), but that might just be me

**Comment by SomeTroglodyte** [comment]
> contents of the ExpanderTab should be

Not "should" - "can", it's just an optional convenience shortcut potentially saving a few lines - that, and the widget can measure the contents earlier if it wants to. [innerTable] is public and so can be manipulated later, even during lifetime. You may (rarely) need more tweaks with container sizing if you fill it after the constructor is done. The Kdoc methinks is clear enough on that.

**Pull request merged**